doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,470 @@
+      using NVMe atomic write command on the WAL
+      - NVMe provides atomicity guarantees for a write command (Atomic Write Unit Power Fail)
+        For example, 512 Kbytes of data can be atomically written at once without fsync()
+      - Small size (object_info_t, snapset, etc.) can be embed
+      - stage 1
+        - if the data is small
+        WAL (written) --> | TxBegin A | Log Entry | TxEnd A | 
+        - if the data is large
+        Data partition (written) --> | Data blocks | 
+      - stage 2
+        - if the data is small
+        No need.
+        - if the data is large
+        Then, append the metadata to WAL.
+        WAL --> | TxBegin A | Log Entry | TxEnd A | 
+
+* Read
+  - Use the cached object metadata to find out the data location
+  - If not cached, need to search WAL after checkpoint and Object meta partition to find the 
+    latest meta data
+
+* Flush (WAL --> Data partition)
+  - Flush WAL entries that have been committed. There are two conditions
+    (1. the size of WAL is close to full, 2. a signal to flush).
+    We can mitigate the overhead of frequent flush via batching processing, but it leads to
+    delaying completion.
+
+
+Crash consistency
+------------------
+* Large case
+  - 1. Crash occurs right after writing Data blocks
+    - Data partition --> | Data blocks |
+    - We don't need to care this case. Data is not alloacted yet in reality. The blocks will be reused.
+  - 2. Crash occurs right after WAL 
+    - Data partition --> | Data blocks |
+    - WAL --> | TxBegin A | Log Entry | TxEnd A |
+    - Write procedure is completed, so there is no data loss or inconsistent state
+
+* Small case
+  - 1. Crash occurs right after writing WAL
+    - WAL --> | TxBegin A | Log Entry| TxEnd A |
+    - All data has been written
+
+
+Comparison
+----------
+- Best case (pre-allocation)
+  - Only need writes on both WAL and Data partition without updating object metadata (for the location).
+- Worst case 
+  - At least three writes are required additionally on WAL, object metadata, and data blocks.
+  - If the flush from WAL to the data parition occurs frequently, b+tree onode structure needs to be update
+    in many times. To minimize such overhead, we can make use of batch processing to minimize the update on the btree
+    (the data related to the object has a locality because it will have the same parent node, so updates can be minimized)
+
+- WAL needs to be flushed if the WAL is close to full or a signal to flush.
+  - The premise behind this design is OSD can manage the latest metadata as a single copy. So,
+    appended entries are not to be read
+- Either best of the worst case does not produce severe I/O amplification (it produce I/Os, but I/O rate is constant) 
+  unlike LSM-tree DB (the proposed design is similar to LSM-tree which has only level-0)
+  
+
+Detailed Design 
+===============
+
+- Onode lookup
+  - Radix tree
```

Comment:
You seem to suggest that this will be a radix tree or a B+ tree in different places, which is it?